experimental/lib/Support/BooleanLogic/BoolExpression.cpp

```diff
@@ -0,0 +1,227 @@
+    s[0] = 'n';
+  } else if (type == ExpressionType::And || type == ExpressionType::Or) {
+    Operator *op = static_cast<Operator *>(this);
+    if (op->left != nullptr)
+      op->left->generateMintermAnd(s, varIndex);
+    if (op->right != nullptr)
+      op->right->generateMintermAnd(s, varIndex);
+  }
+}
+
+void BoolExpression::generateMintermsOr(
+    int numOfVariables, const std::map<std::string, int> &varIndex,
+    std::set<std::string> &minterms) {
+  if (type == ExpressionType::Or) {
+    Operator *op = static_cast<Operator *>(this);
+    op->left->generateMintermsOr(numOfVariables, varIndex, minterms);
+    op->right->generateMintermsOr(numOfVariables, varIndex, minterms);
+  } else if (type == ExpressionType::One) { // 1 + exp = 1;
+    std::string s(numOfVariables, 'd');
+    minterms.insert(s);
+  } else if (type == ExpressionType::Zero) { // 0 + exp = exp
+    return;
+  } else {
+    std::string s(numOfVariables, 'd'); // initializing s
+    generateMintermAnd(s, varIndex);
+    if (s.find('n') ==
+        std::string::npos) // no null in the minterm -> minterm is valid
+      minterms.insert(s);
+  }
+}
+
+// 1- generate all the minterms with don't cars
+// 2- replace the don't cares wit 0s and 1s
+// 3- loop over all rows in the truth table and check wether it corresponds to a
+// minterm or not
+std::set<std::string> BoolExpression::generateTruthTable() {
+  std::set<std::string> variables = BoolExpression::getVariables();
+  int numOfVariables = variables.size();
+  std::map<std::string, int> varIndex;
+  int index = 0;
+  for (const std::string &s : variables) {
+    varIndex[s] = index++;
+  }
+  // generate the minterms
+  std::set<std::string> minterms;
+  generateMintermsOr(numOfVariables, varIndex, minterms);
+  std::set<std::string> mintermsWithoutDontCares = replaceDontCares(minterms);
+  // generate the truth table
+  std::string s(numOfVariables, 'd');
+  std::set<std::string> r = {s};
+  std::set<std::string> rows = replaceDontCares(r);
+  std::set<std::string> table;
+  for (const std::string &row : rows) {
+    // if the row corresponds to the minterm, then f is 1
+    if (mintermsWithoutDontCares.find(row) != mintermsWithoutDontCares.end())
+      table.insert((row + "1"));
+    // else f is 0
+    else
+      table.insert((row + "0"));
+  }
+  return table;
+}
+
+//--------Printing BoolExpresssion--------
+static constexpr unsigned COUNT = 10;
+
+// Inspired from https://www.geeksforgeeks.org/print-binary-tree-2-dimensions/
+void BoolExpression::print(int space) {
+  // Increase distance between levels
+  space += COUNT;
+
+  // Process right child first
+  if (type == ExpressionType::Or || type == ExpressionType::And) {
+    Operator *op = static_cast<Operator *>(this);
+    op->right->print(space);
+  }
+  // Print current node after space
+  // count
+  llvm::outs() << "\n";
+  for (int i = COUNT; i < space; i++)
+    llvm::outs() << " ";
+
+  if (type == ExpressionType::Variable) {
```

Comment:
Same comment as somewhere above, use a `switch` if mean to check for all possible `type` values.